# Foundations of Computer Science - Project (2024-2025)

## Elena Maggiore 880202
## Rebecca Micol Finzi 882523

For this project, we will work on the **NYC TRIP FARE** dataset. First we start off with some preliminary steps: we import the **pandas** library, ignore eventual warnings, read the dataset and visualize its first rows.

In [3]:
import pandas as pd #importing the pandas library.

import warnings #importing the warnings library.
warnings.filterwarnings('ignore') #since later some warnings will occur, we choose to ignore them to improve results visualization.

data=pd.read_csv("C:/Users/37350/Desktop/Progetti Scuola/MAGISTRALE/PRIMO ANNO/Esame Foundations of Computer Science/Project/data.csv", low_memory=False) #reading the dataset, low_memory=False is used due to the high dimensionality of the dataset.
data.head(5) #visualizing the first 5 rows.

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0       1.0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0   
1       1.0  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0   
2       1.0  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0   
3       1.0  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0   
4       2.0  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            1.2         1.0                  N           238           239   
1            1.2         1.0                  N           239           238   
2            0.6         1.0                  N           238           238   
3            0.8         1.0                  N           238           151   
4            0.0         1.0                  N           193           193   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0           1.0          6.0    3.0      0.5        1.47           0.0   
1           1.0          7.0    3.0      0.5        1.50           0.0   
2           1.0          6.0    3.0      0.5        1.00           0.0   
3           1.0          5.5    0.5      0.5        1.36           0.0   
4           2.0          3.5    0.5      0.5        0.00           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  
0                    0.3         11.27                   2.5  
1                    0.3         12.30                   2.5  
2                    0.3         10.80                   2.5  
3                    0.3          8.16                   0.0  
4                    0.3          4.80                   0.0

The assignment then states:

*"You can skip the **store_and_fwd_flag** column, but it’s a bonus point if you can manage it correctly."*

so, we first take a look at the **store_and_fwd_flag** column and its possible levels.

In [7]:
data["store_and_fwd_flag"] #visualizing the entire column.

values=data["store_and_fwd_flag"].value_counts(dropna=False) #obtaining the levels of the variable, counting NaNs.
print(values)

store_and_fwd_flag
N      6271447
Y        68120
NaN      65441
Name: count, dtype: int64


We see how the variable **"store_and_fwd_flag"** values 'y' 68120 times, 'n' 6271447 times, and NaN 65441 times.
Noticing how 'y' could resemble the word 'yes' and similarly 'n' could mean 'no', we then decide to convert the variable to the Boolean type, where 'y' means **True** and 'n' means **False**. 

In [9]:
data["store_and_fwd_flag"]=data["store_and_fwd_flag"].replace("N", False).replace("Y", True)
data.head(2)   #let's see the data now

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0       1.0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0   
1       1.0  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            1.2         1.0              False           238           239   
1            1.2         1.0              False           239           238   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0           1.0          6.0    3.0      0.5        1.47           0.0   
1           1.0          7.0    3.0      0.5        1.50           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  
0                    0.3         11.27                   2.5  
1                    0.3         12.30                   2.5

Then we focus on the NaNs. Since the five columns mentioned before all have 65441 NaNs, we ask ourselves if those NaNs are on the same rows; if that's the case, we can remove them altogether.

In [11]:
(data["store_and_fwd_flag"].isna()&data["VendorID"].isna()&data["passenger_count"].isna()&data["RatecodeID"].isna()&data["payment_type"].isna()).sum()

65441

There are 65441 rows where all the variables in exam value NaN, so we decide to drop them.

In [13]:
data=data.dropna()

Now we see how there dataset contains no NaNs anymore.

In [15]:
data.isna().sum()

VendorID                 0
tpep_pickup_datetime     0
tpep_dropoff_datetime    0
passenger_count          0
trip_distance            0
RatecodeID               0
store_and_fwd_flag       0
PULocationID             0
DOLocationID             0
payment_type             0
fare_amount              0
extra                    0
mta_tax                  0
tip_amount               0
tolls_amount             0
improvement_surcharge    0
total_amount             0
congestion_surcharge     0
dtype: int64

After removing 65441 rows, we should have 6405008-65441=6339567 rows left.

In [17]:
data.shape[0] #6339567 rows returning.

6339567

Since some exercises will later require us to work on the dataset with NaNs, we save another dataset that still contains them.

In [19]:
data_with_na=pd.read_csv("C:/Users/37350/Desktop/Progetti Scuola/MAGISTRALE/PRIMO ANNO/Esame Foundations of Computer Science/Project/data.csv", low_memory=False)

- ### 1) Extract all trips with *trip_distance* larger than 50.

We pick the rows where *trip_distance* is higher than 50.

In [22]:
data[data["trip_distance"]>50] #377 trips returned.

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
23842         2.0  2020-01-01 01:53:07   2020-01-01 03:54:41              1.0   
39013         2.0  2020-01-01 02:05:07   2020-01-01 03:03:10              1.0   
41620         1.0  2020-01-01 03:05:54   2020-01-01 04:16:26              1.0   
58262         2.0  2020-01-01 05:36:12   2020-01-01 06:40:06              1.0   
63024         2.0  2020-01-01 07:40:30   2020-01-01 08:40:01              1.0   
...           ...                  ...                   ...              ...   
6287089       2.0  2020-01-31 19:43:38   2020-01-31 21:12:59              1.0   
6322321       2.0  2020-01-31 22:06:26   2020-01-31 23:40:19              3.0   
6326169       2.0  2020-01-31 22:47:26   2020-01-31 23:49:14              1.0   
6331181       2.0  2020-01-31 23:45:36   2020-02-01 01:00:25              5.0   
6333801       2.0  2020-01-31 23:24:16   2020-02-01 01:32:56              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
23842            52.30         5.0              False           262   
39013            51.23         5.0              False           264   
41620            53.80         5.0              False           132   
58262            55.23         5.0              False           132   
63024            54.19         5.0              False           132   
...                ...         ...                ...           ...   
6287089          53.17         5.0              False           138   
6322321          56.49         5.0              False           170   
6326169          51.83         5.0              False           132   
6331181          57.99         4.0              False           107   
6333801          52.97         4.0              False           264   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
23842             265           1.0        300.0    0.0      0.0       61.78   
39013             264           1.0        329.0    0.0      0.5      100.78   
41620             265           1.0        250.0    0.0      0.0       53.35   
58262             265           2.0        170.0    0.0      0.5        0.00   
63024             265           1.0        230.0    0.0      0.0        0.00   
...               ...           ...          ...    ...      ...         ...   
6287089           264           2.0        180.0    2.5      0.5        0.00   
6322321           265           1.0        100.0    0.0      0.0       25.70   
6326169           265           1.0        220.0    0.0      0.5       48.96   
6331181           265           1.0        245.0    0.5      0.5       38.24   
6333801           265           1.0        227.0    0.5      0.5       46.16   

         tolls_amount  improvement_surcharge  total_amount  \
23842            6.12                    0.3        370.70   
39013            6.12                    0.3        436.70   
41620           16.62                    0.3        320.27   
58262           18.26                    0.3        189.06   
63024           12.24                    0.3        242.54   
...               ...                    ...           ...   
6287089          0.00                    0.3        183.30   
6322321          0.00                    0.3        128.50   
6326169         23.99                    0.3        293.75   
6331181          6.12                    0.3        293.16   
6333801          0.00                    0.3        276.96   

         congestion_surcharge  
23842                     2.5  
39013                     0.0  
41620                     0.0  
58262                     0.0  
63024                     0.0  
...                       ...  
6287089                   0.0  
6322321                   2.5  
6326169                   0.0  
6331181                   2.5  
6333801                   2.5  

[377 rows x 18 columns]

- ### 2) Extract all trips where *payment_type* is missing.

For this exercise we use the dataset we copied, the one that still contains NaNs.
To extract all trips where *payment_type* is missing can be done with the same function *.isna()* we used before.

In [25]:
data_with_na[data_with_na["payment_type"].isna()] #65441 trips returned.

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
6339567       NaN  2020-01-01 08:51:00   2020-01-01 09:19:00              NaN   
6339568       NaN  2020-01-01 08:38:43   2020-01-01 08:51:08              NaN   
6339569       NaN  2020-01-01 08:27:00   2020-01-01 08:32:00              NaN   
6339570       NaN  2020-01-01 08:46:00   2020-01-01 08:57:00              NaN   
6339571       NaN  2020-01-01 08:21:00   2020-01-01 08:38:00              NaN   
...           ...                  ...                   ...              ...   
6405003       NaN  2020-01-31 22:51:00   2020-01-31 23:22:00              NaN   
6405004       NaN  2020-01-31 22:10:00   2020-01-31 23:26:00              NaN   
6405005       NaN  2020-01-31 22:50:07   2020-01-31 23:17:57              NaN   
6405006       NaN  2020-01-31 22:25:53   2020-01-31 22:48:32              NaN   
6405007       NaN  2020-01-31 22:44:00   2020-01-31 23:06:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
6339567          13.69         NaN                NaN           136   
6339568           3.42         NaN                NaN           121   
6339569           2.20         NaN                NaN           197   
6339570           0.84         NaN                NaN           262   
6339571           7.24         NaN                NaN            45   
...                ...         ...                ...           ...   
6405003           3.24         NaN                NaN           237   
6405004          22.13         NaN                NaN           259   
6405005          10.51         NaN                NaN           137   
6405006           5.49         NaN                NaN            50   
6405007          11.60         NaN                NaN           179   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
6339567           232           NaN        51.05   2.75      0.5         0.0   
6339568             9           NaN        27.06   2.75      0.0         0.0   
6339569           216           NaN        24.36   2.75      0.5         0.0   
6339570           236           NaN        26.08   2.75      0.5         0.0   
6339571           142           NaN        25.28   2.75      0.5         0.0   
...               ...           ...          ...    ...      ...         ...   
6405003           234           NaN        17.59   2.75      0.5         0.0   
6405004            45           NaN        46.67   2.75      0.5         0.0   
6405005           169           NaN        48.85   2.75      0.0         0.0   
6405006            42           NaN        27.17   2.75      0.0         0.0   
6405007           205           NaN        54.56   2.75      0.5         0.0   

         tolls_amount  improvement_surcharge  total_amount  \
6339567          0.00                    0.3         54.60   
6339568          0.00                    0.3         30.11   
6339569          0.00                    0.3         27.91   
6339570          0.00                    0.3         29.63   
6339571          0.00                    0.3         28.83   
...               ...                    ...           ...   
6405003          0.00                    0.3         21.14   
6405004         12.24                    0.3         62.46   
6405005          0.00                    0.3         51.90   
6405006          0.00                    0.3         30.22   
6405007          0.00                    0.3         58.11   

         congestion_surcharge  
6339567                   0.0  
6339568                   0.0  
6339569                   0.0  
6339570                   0.0  
6339571                   0.0  
...                       ...  
6405003                   0.0  
6405004                   0.0  
6405005                   0.0  
6405006                   0.0  
6405007                   0.0  

[65441 rows x 18 columns]

- ### 3) For each *(PULocationID, DOLocationID)* pair, determine the number of trips.

We get back to using the dataset without NaNs.
This exercise can be solved by using the *.groupby* function.

In [28]:
data.groupby(["PULocationID", "DOLocationID"]).size() #grouping by the pair PULocationID, DOLocationID. 26456 pairs returned.
d3=data.groupby(["PULocationID", "DOLocationID"]) #saving the results.

Using the function *get_group*, we can visualize all trips in a single group.

In [30]:
d3.get_group((1,1)) #example with all trips of group (1,1). 636 trips returned.

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
57669         2.0  2020-01-01 04:41:47   2020-01-01 04:42:22              4.0   
61345         2.0  2020-01-01 06:54:53   2020-01-01 06:55:13              1.0   
61377         1.0  2020-01-01 06:57:17   2020-01-01 06:58:01              4.0   
61426         2.0  2020-01-01 06:20:55   2020-01-01 06:21:09              2.0   
62600         2.0  2020-01-01 06:53:39   2020-01-01 06:53:49              1.0   
...           ...                  ...                   ...              ...   
6252251       1.0  2020-01-31 17:28:23   2020-01-31 17:28:52              1.0   
6261853       2.0  2020-01-31 18:01:03   2020-01-31 18:01:09              1.0   
6277348       1.0  2020-01-31 19:21:56   2020-01-31 19:23:10              1.0   
6278938       1.0  2020-01-31 19:45:17   2020-01-31 19:52:05              1.0   
6287641       2.0  2020-01-31 19:13:22   2020-01-31 19:13:41              2.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
57669              0.0         5.0              False             1   
61345              0.0         5.0              False             1   
61377              0.0         5.0              False             1   
61426              0.0         5.0              False             1   
62600              0.0         5.0              False             1   
...                ...         ...                ...           ...   
6252251            0.0         5.0              False             1   
6261853            0.0         5.0              False             1   
6277348           10.3         5.0              False             1   
6278938            0.0         5.0              False             1   
6287641            0.0         5.0              False             1   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
57669               1           1.0         89.0    0.0      0.5        5.00   
61345               1           1.0         96.0    0.0      0.0        5.08   
61377               1           1.0         84.0    0.0      0.0        5.00   
61426               1           1.0        150.0    0.0      0.5        0.00   
62600               1           1.0         60.0    0.0      0.0        0.00   
...               ...           ...          ...    ...      ...         ...   
6252251             1           1.0         85.0    0.0      0.0       17.05   
6261853             1           1.0         90.0    0.0      0.0       18.06   
6277348             1           1.0        100.0    0.0      0.0        0.01   
6278938             1           1.0        150.0    0.0      0.0        0.00   
6287641             1           1.0        130.0    0.0      0.5       10.00   

         tolls_amount  improvement_surcharge  total_amount  \
57669            0.00                    0.3         94.80   
61345           14.50                    0.3        115.88   
61377            0.00                    0.3         89.30   
61426           29.00                    0.3        179.80   
62600            0.00                    0.3         60.30   
...               ...                    ...           ...   
6252251          0.00                    0.3        102.35   
6261853          0.00                    0.3        108.36   
6277348         13.75                    0.3        114.06   
6278938         15.00                    0.3        165.30   
6287641         27.50                    0.3        168.30   

         congestion_surcharge  
57669                     0.0  
61345                     0.0  
61377                     0.0  
61426                     0.0  
62600                     0.0  
...                       ...  
6252251                   0.0  
6261853                   0.0  
6277348                   0.0  
6278938                   0.0  
6287641                   0.0  

[636 rows x 18 columns]

- ### 4) Save all rows with missing *VendorID*, *passenger_count*, *store_and_fwd_flag*, *payment_type* in a new dataframe called **bad**, and remove those rows from the original dataframe.

We've already removed NaNs from the original dataframe, so now we do it again on the **data_with_na** dataframe, copying the rows that contain missing values on the dataframe **bad**.

In [33]:
bad=data_with_na[data_with_na[["VendorID", "passenger_count", "store_and_fwd_flag", "payment_type"]].isna().any(axis=1)]
bad #dimensions: 65441 rows x 18 columns.

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
6339567       NaN  2020-01-01 08:51:00   2020-01-01 09:19:00              NaN   
6339568       NaN  2020-01-01 08:38:43   2020-01-01 08:51:08              NaN   
6339569       NaN  2020-01-01 08:27:00   2020-01-01 08:32:00              NaN   
6339570       NaN  2020-01-01 08:46:00   2020-01-01 08:57:00              NaN   
6339571       NaN  2020-01-01 08:21:00   2020-01-01 08:38:00              NaN   
...           ...                  ...                   ...              ...   
6405003       NaN  2020-01-31 22:51:00   2020-01-31 23:22:00              NaN   
6405004       NaN  2020-01-31 22:10:00   2020-01-31 23:26:00              NaN   
6405005       NaN  2020-01-31 22:50:07   2020-01-31 23:17:57              NaN   
6405006       NaN  2020-01-31 22:25:53   2020-01-31 22:48:32              NaN   
6405007       NaN  2020-01-31 22:44:00   2020-01-31 23:06:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
6339567          13.69         NaN                NaN           136   
6339568           3.42         NaN                NaN           121   
6339569           2.20         NaN                NaN           197   
6339570           0.84         NaN                NaN           262   
6339571           7.24         NaN                NaN            45   
...                ...         ...                ...           ...   
6405003           3.24         NaN                NaN           237   
6405004          22.13         NaN                NaN           259   
6405005          10.51         NaN                NaN           137   
6405006           5.49         NaN                NaN            50   
6405007          11.60         NaN                NaN           179   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
6339567           232           NaN        51.05   2.75      0.5         0.0   
6339568             9           NaN        27.06   2.75      0.0         0.0   
6339569           216           NaN        24.36   2.75      0.5         0.0   
6339570           236           NaN        26.08   2.75      0.5         0.0   
6339571           142           NaN        25.28   2.75      0.5         0.0   
...               ...           ...          ...    ...      ...         ...   
6405003           234           NaN        17.59   2.75      0.5         0.0   
6405004            45           NaN        46.67   2.75      0.5         0.0   
6405005           169           NaN        48.85   2.75      0.0         0.0   
6405006            42           NaN        27.17   2.75      0.0         0.0   
6405007           205           NaN        54.56   2.75      0.5         0.0   

         tolls_amount  improvement_surcharge  total_amount  \
6339567          0.00                    0.3         54.60   
6339568          0.00                    0.3         30.11   
6339569          0.00                    0.3         27.91   
6339570          0.00                    0.3         29.63   
6339571          0.00                    0.3         28.83   
...               ...                    ...           ...   
6405003          0.00                    0.3         21.14   
6405004         12.24                    0.3         62.46   
6405005          0.00                    0.3         51.90   
6405006          0.00                    0.3         30.22   
6405007          0.00                    0.3         58.11   

         congestion_surcharge  
6339567                   0.0  
6339568                   0.0  
6339569                   0.0  
6339570                   0.0  
6339571                   0.0  
...                       ...  
6405003                   0.0  
6405004                   0.0  
6405005                   0.0  
6405006                   0.0  
6405007                   0.0  

[65441 rows x 18 columns]

Then we remove from **data_with_na** the rows contained in **bad**, referring to their indexes. We save the results in a new dataframe called **data_without_na** and verify that no NaNs are there.

In [35]:
bad.index #indexes of the rows that contain NaNs.
data_without_na=data_with_na.drop((bad.index), axis=0) #removing those rows and saving the results in a new dataframe.
data_without_na #6339567 rows returned.
data_without_na.isnull().sum() #this dataframe contains no NaNs.

VendorID                 0
tpep_pickup_datetime     0
tpep_dropoff_datetime    0
passenger_count          0
trip_distance            0
RatecodeID               0
store_and_fwd_flag       0
PULocationID             0
DOLocationID             0
payment_type             0
fare_amount              0
extra                    0
mta_tax                  0
tip_amount               0
tolls_amount             0
improvement_surcharge    0
total_amount             0
congestion_surcharge     0
dtype: int64

- ### 5) Add a duration column storing how long each trip has taken (use *tpep_pickup_datetime*, *tpep_dropoff_datetime*).

First we convert both columns to the *datetime* format, then we just compute the difference and save it in a new column called *duration*.

In [38]:
data['tpep_pickup_datetime']=pd.to_datetime(data['tpep_pickup_datetime']) #converting the variables to the datetime format.
data['tpep_dropoff_datetime']=pd.to_datetime(data['tpep_dropoff_datetime'])

data["duration"]=data["tpep_dropoff_datetime"]-data["tpep_pickup_datetime"] #computing the difference.
data.head(1) #first row's trip duration is 0 days 00:04:48.

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0       1.0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            1.2         1.0              False           238           239   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0           1.0          6.0    3.0      0.5        1.47           0.0   

   improvement_surcharge  total_amount  congestion_surcharge        duration  
0                    0.3         11.27                   2.5 0 days 00:04:48

- ### 6) For each pickup location, determine how many trips have started there.

Initially we group the data by pickup location.

In [41]:
data.groupby(["PULocationID"]).size() #260 groups returned.

PULocationID
1         753
2           3
3          70
4        9902
5          39
        ...  
261     34229
262     85591
263    123997
264     43779
265      3090
Length: 260, dtype: int64

By looking at the indexes, we see that *PULocationID* goes up to 265, but we only have 260 groups; that is supposedly because some locations could be dropoff locations only, and so they may not appear in the *PULocationID* column. What are their *LocationID*s?

In [43]:
pickup_locations=data.groupby(["PULocationID"]).size() #saving all pickup locations.
dropoff_only=set(range(1, 266))-set(pickup_locations.index) #supposing we have 265 different LocationIDs, which ones of them are not in pickup_locations?
dropoff_only #LocationID 27, 103, 104, 110, 199 returning.

{27, 103, 104, 110, 199}

- ### 7) Cluster the pickup time of the day into 30-minute intervals (e.g. from 02:00 to 02:30).

We have already converted the *tpep_pickup_datetime* and *tpep_pickup_datetime* variables to *datetime*, so we can pick both time and date with the *.dt* functions and save them in new columns. Specifically we're going to need the *pickup_time* variable for later. 

In [46]:
data["pickup_date"]=data['tpep_pickup_datetime'].dt.date #saving the pickup date in a new column.
data["pickup_time"]=data['tpep_pickup_datetime'].dt.time #saving the pickup time in a new column.

data["dropoff_date"]=data['tpep_dropoff_datetime'].dt.date #doing the same for dropoff date and time.
data["dropoff_time"]=data['tpep_dropoff_datetime'].dt.time

data.head(1) #for the first trip, pickup_date is 2020-01-01, pickup_time is 00:28:15, dropoff_date is 2020-01-01, dropoff_time is 00:33:03.

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0       1.0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            1.2         1.0              False           238           239   

   payment_type  ...  tip_amount  tolls_amount  improvement_surcharge  \
0           1.0  ...        1.47           0.0                    0.3   

   total_amount  congestion_surcharge        duration  pickup_date  \
0         11.27                   2.5 0 days 00:04:48   2020-01-01   

   pickup_time dropoff_date dropoff_time  
0     00:28:15   2020-01-01     00:33:03  

[1 rows x 23 columns]

Now we define the function *interval* that, for each *pickup_time*, returns the 30-minute interval it's in. To make it work then, we have to convert *pickup_time* in the *datetime H:M:S* format.

In [48]:
def interval(time):
    m=time.minute #minutes.
    h=time.hour #hours.
    
    if m<30: #case where the time falls into the first half of the hour.
        return f"{h:02d}:00-{h:02d}:30" #writing the interval.       
    else: #case where the time falls into the second half of the hour.
        next_h=(h+1)%24 #getting the next hour, (remembering that after 23 comes 0).
        return f"{h:02d}:30-{next_h:02d}:00" #writing the interval.

data['pickup_time']=pd.to_datetime(data['pickup_time'], format='%H:%M:%S').dt.time #converting pickup_time in the datetime H:M:S format.
data['interval']=data['pickup_time'].apply(interval) #applying the function on the dataframe.
data.head(1) #the first trip falls into the 00:00 - 00:30 interval.

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0       1.0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            1.2         1.0              False           238           239   

   payment_type  ...  tolls_amount  improvement_surcharge  total_amount  \
0           1.0  ...           0.0                    0.3         11.27   

   congestion_surcharge        duration  pickup_date  pickup_time  \
0                   2.5 0 days 00:04:48   2020-01-01     00:28:15   

   dropoff_date dropoff_time     interval  
0    2020-01-01     00:33:03  00:00-00:30  

[1 rows x 24 columns]

- ### 8) For each interval, determine the average number of passengers and the average fare amount.

We group the data by *interval* and then compute the mean of the chosen variables.

In [51]:
data.groupby("interval")[["passenger_count","fare_amount"]].mean() #for the 00:00 - 00:30 interval, passenger_count	average value is 1.572848 and fare_amount average value is 13.526433.

passenger_count  fare_amount
interval                                 
00:00-00:30         1.572848    13.526433
00:30-01:00         1.584345    13.214132
01:00-01:30         1.578933    12.699554
01:30-02:00         1.589182    12.265997
02:00-02:30         1.587479    12.089669
02:30-03:00         1.587687    12.041626
03:00-03:30         1.582064    12.500846
03:30-04:00         1.585838    13.094785
04:00-04:30         1.580261    14.192685
04:30-05:00         1.515886    16.409774
05:00-05:30         1.483321    18.633908
05:30-06:00         1.414368    15.692556
06:00-06:30         1.393482    14.578730
06:30-07:00         1.383740    12.321038
07:00-07:30         1.415469    12.268222
07:30-08:00         1.449612    11.778843
08:00-08:30         1.458988    11.953842
08:30-09:00         1.444167    11.874444
09:00-09:30         1.453355    12.072809
09:30-10:00         1.469457    11.884284
10:00-10:30         1.486685    12.053572
10:30-11:00         1.494183    11.842533
11:00-11:30         1.502240    11.888702
11:30-12:00         1.500619    11.638112
12:00-12:30         1.499564    11.905048
12:30-13:00         1.499128    11.981302
13:00-13:30         1.506846    12.363283
13:30-14:00         1.514286    12.562459
14:00-14:30         1.511807    12.975308
14:30-15:00         1.509193    12.783314
15:00-15:30         1.525653    12.886364
15:30-16:00         1.522081    12.889025
16:00-16:30         1.515844    13.282861
16:30-17:00         1.512844    12.947138
17:00-17:30         1.514176    12.686836
17:30-18:00         1.523290    12.226527
18:00-18:30         1.518246    12.157125
18:30-19:00         1.525598    11.908637
19:00-19:30         1.525397    11.871145
19:30-20:00         1.529090    12.006640
20:00-20:30         1.530152    12.502959
20:30-21:00         1.539479    12.707428
21:00-21:30         1.551734    12.744031
21:30-22:00         1.564563    12.559674
22:00-22:30         1.565824    12.805710
22:30-23:00         1.571608    13.170497
23:00-23:30         1.569022    13.253366
23:30-00:00         1.570120    13.206596

- ### 9) For each payment type and each interval, determine the average fare amount.

This time, we group by both *payment_type* and *interval* before computing the mean of the variable *fare_amount*.

In [54]:
data.groupby(["payment_type","interval"])["fare_amount"].mean() #for payment_type 1.0 and interval 00:00 - 00:30, the average fare amount is 13.869142.

payment_type  interval   
1.0           00:00-00:30    13.869142
              00:30-01:00    13.472232
              01:00-01:30    12.824603
              01:30-02:00    12.357974
              02:00-02:30    12.008589
                               ...    
4.0           22:00-22:30     1.533326
              22:30-23:00    -0.787090
              23:00-23:30    -0.351277
              23:30-00:00    -2.748432
5.0           17:30-18:00     0.000000
Name: fare_amount, Length: 193, dtype: float64

- ### 10) For each payment type, determine the interval when the average fare amount is maximum.

First we save the data after grouping it by *payment_type* and *interval* and obtaining for each group the average fare amount, then we group the data again by *payment_type* and for each group we extract the index of the maximum fare amount average.

In [57]:
grouped_data=data.groupby(["payment_type", "interval"])["fare_amount"].mean() #computing the average fare amount for each payment_type, fare_amount pair.
grouped_data.groupby("payment_type").idxmax() #for payment_type 1.0, the interval with the highest average fare amount is 05:00 - 05:30.

payment_type
1.0    (1.0, 05:00-05:30)
2.0    (2.0, 05:00-05:30)
3.0    (3.0, 07:00-07:30)
4.0    (4.0, 05:00-05:30)
5.0    (5.0, 17:30-18:00)
Name: fare_amount, dtype: object

- ### 11) For each payment type, determine the interval when the overall ratio between the tip and the fare amounts is maximum.

We group the data by *payment_type* and *interval*, then we save in the **grouped1** and **grouped2** datasets respectively the total *tip_amount* and the total *fare_amount*.

In [60]:
grouped1=data.groupby(["payment_type", "interval"])["tip_amount"].sum() #dataframe with total tip_amount.
grouped2=data.groupby(["payment_type", "interval"])["fare_amount"].sum() #dataframe with total fare_amount.

Then we compute the ratio between each row of the two dataset, using the operator *.div* since we want to remove NaNs that will be generated by dividing by 0.

In [62]:
grouped_data=grouped1.div(grouped2, fill_value=0) #fill_value=0 replaces all NaNs generated with 0.

And we find, for each payment type, the interval when the overall ratio is maximum.

In [64]:
grouped_data.groupby("payment_type").idxmax() #for the first payment_type, the interval when the overall ratio is maximum is 18:30 - 19:00.

payment_type
1.0    (1.0, 18:30-19:00)
2.0    (2.0, 05:00-05:30)
3.0    (3.0, 21:00-21:30)
4.0    (4.0, 13:00-13:30)
5.0                   NaN
dtype: object

- ### 12) Find the location with the highest average fare amount.

We didn't understand if this question asked for both the pickup and dropoff locations with the highest average fare amount, or if we should have merged the pickup and dropoff locations and their fare amounts. We chose to do the latter.

First, we compute the average fare_amount for both the pickup and dropoff locations.

In [68]:
loc1=data.groupby("PULocationID")["fare_amount"].mean() #pickup locations fare_amount averages.
loc2=data.groupby("DOLocationID")["fare_amount"].mean() #dropoff locations fare_amount averages.

loc1=pd.DataFrame(loc1) #they need to be dataframes to merge.
loc2=pd.DataFrame(loc2)

Now we merge them in a single dataframe that contains all locations and their average fare amount as pickup location and as dropoff location.

In [70]:
locations=loc1.merge(loc2, how="outer", left_on="PULocationID", right_on="DOLocationID", right_index= True) #using outer join to include all locations, and right_index after checking that every location has been a dropoff location at least once.
locations=locations.rename(columns={'PULocationID': 'Location','fare_amount_x': 'PU_fare_amount','fare_amount_y': 'DO_fare_amount'}) #renaming the columns.
locations.head(1) #first location has a fare_amount average of 82.242802 as pickup location and of 70.762930 as dropoff location.

Location  PU_fare_amount  DO_fare_amount
1.0         1       82.242802        70.76293

Before computing the mean between the two cases, we must address the NaNs (generated when a location appeared as dropoff location, but not as pickup location); by replacing them with the value that location assumes in the other column, we ensure that the mean remains the same.

In [72]:
locations['PU_fare_amount']=locations['PU_fare_amount'].fillna(locations['DO_fare_amount']) #filling the NaNs of with the value of the other column.
locations.set_index('Location', inplace=True) #setting the indexes to match the LocationID.
locations.head(1) #now the index of the first location is 1.

PU_fare_amount  DO_fare_amount
Location                                
1              82.242802        70.76293

After removing NaNs and normalizing indexes, we can finally compute the mean between the two columns, save it in a new dataframe and extract the location with the highest average fare amount.

In [74]:
locations_mean=((locations["PU_fare_amount"]+locations["DO_fare_amount"])/2) #computing the mean between the two columns.
locations_mean.idxmax() #location 204 returned.

204

- ### 13) Build a new dataframe (called **common**) where, for each pickup location we keep all trips to the 5 most common destinations (i.e. each pickup location can have different common destinations).

First we save all *PULocationID* and *DOLocationID* combinations in a dataframe and count how many trips are there for each of them.

In [77]:
combinations=data.groupby(["PULocationID","DOLocationID"]).count()["VendorID"] #grouping data on PULocationID and DOLocationID.
combinations=pd.DataFrame(combinations) #converting it to a dataframe.
combinations.head(15) #for PULocationID 1 and DOLocationID 1, 636 trips returning.

VendorID
PULocationID DOLocationID          
1            1                  636
             50                   1
             68                   1
             138                  2
             140                  1
             148                  1
             211                  1
             231                  1
             264                105
             265                  4
2            56                   1
             95                   1
             201                  1
3            3                    5
             4                    1

Then we sort the data from the combination with the highest number of trips and descreasingly, and group it again by pickup location visualizing only the first five rows. 

In [79]:
combinations=combinations.sort_values(["PULocationID","VendorID"], ascending=False) #sorting all combinations.
combinations=combinations.groupby("PULocationID").head(5) #saving only the first 5 rows per group.

In [80]:
combinations.head(5) #for group 265, the 5 most common destinations are location 265, 264, 1, 48, 132.

VendorID
PULocationID DOLocationID          
265          265               2496
             264                317
             1                   76
             48                  18
             132                  8

Now we know, for each pickup location, the 5 most common destinations. We can use *.reset_index* to have *PULocationID* and *DOLocationID* as columns instead of indexes, and then create the **common** dataset by keeping only the rows of the original dataframe where *PULocationID* and *DULocationID* correspond.

In [82]:
combinations=combinations.reset_index()[['PULocationID', 'DOLocationID']] #resetting the index, making PULocationID and DOLocationID columns instead.
common=data.merge(combinations, on=['PULocationID', 'DOLocationID'], how='inner') #merging the datasets with an inner join removes all rows where the combination of PULocationID, DOLocationID is not in combinations.  
common.head(1) #the first trip on the dataset common starts at location 238 at 00.28.15 and ends at location 239 at 00.33.03.

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0       1.0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            1.2         1.0              False           238           239   

   payment_type  ...  tolls_amount  improvement_surcharge  total_amount  \
0           1.0  ...           0.0                    0.3         11.27   

   congestion_surcharge        duration  pickup_date  pickup_time  \
0                   2.5 0 days 00:04:48   2020-01-01     00:28:15   

   dropoff_date dropoff_time     interval  
0    2020-01-01     00:33:03  00:00-00:30  

[1 rows x 24 columns]

- ### 14) On the **common** dataframe, for each payment type and each interval, determine the average fare amount.

We follow the same process we used for exercise 9.

In [85]:
common.groupby(["payment_type","interval"])["fare_amount"].mean() #for payment type 1.0 and interval 00:00 - 00:30, the average fare_amount is 8.588957.

payment_type  interval   
1.0           00:00-00:30    8.588957
              00:30-01:00    8.690681
              01:00-01:30    8.496143
              01:30-02:00    8.026808
              02:00-02:30    7.945909
                               ...   
4.0           22:00-22:30    1.300119
              22:30-23:00   -0.563944
              23:00-23:30   -2.568571
              23:30-00:00   -4.409173
5.0           17:30-18:00    0.000000
Name: fare_amount, Length: 193, dtype: float64

- ### 15) Compute the difference of the average fare amount computed in the previous point with those computed at point 9.

We can simply obtain again the result we got in point 9 and in point 14, and then compute the difference.

In [88]:
p9=data.groupby(["payment_type","interval"])["fare_amount"].mean() #copying from point 9.
p14=common.groupby(["payment_type","interval"])["fare_amount"].mean() #copying from point 14.

p15=p9-p14 #computing the difference.
p15.head(1) #for payment type 1 and interval 00:00 - 00:30 the difference is 5.280185.

payment_type  interval   
1.0           00:00-00:30    5.280185
Name: fare_amount, dtype: float64

- ### 16) Compute the ratio between the differences computed in the previous point and those computed in point 9.
    #### Note: you have to compute a ratio for each pair (*payment type*, *interval*).

We already have the results of point 9 and point 15 saved, so we just compute the ratio. We use the *.div* operator, so we can replace NaNs with "Inf" since they're generated when dividing by 0.

In [91]:
p16=p9.div(p15, fill_value='Inf') #computing the ratio.
p16.head(1) #for payment type 1 and interval 00:00 - 00:30, the ratio is 2.62664.

payment_type  interval   
1.0           00:00-00:30    2.62664
Name: fare_amount, dtype: float64

- ### 17) Build chains of trips.
#### Two trips are consecutive in a chain if:
- (a) they have the same *VendorID*.
- (b) the pickup location of the second trip is also the dropoff location of the first trip.
- (c) the pickup time of the second trip is after the dropoff time of the first trip.
- (d) the pickup time of the second trip is at most 2 minutes later than the dropoff time of the first trip.

Since obtaining every possible chain would mean using a recursive algorithm (which has complexity **O(n!)** and so is computationally impossible to execute), we decided to change our approach; we won't try to get every possible chain, but rather for each trip we'll add the column *chains_to* to show every possible other trip it can be chained to.

First, we'll try and clean the dataset, removing all trips with *trip_distance=0* or *total_amount<0*.

In [95]:
data.shape[0] #the dataset is made of 6339567 rows.

data['trip_distance']=data['trip_distance'].astype(float) #converting trip_distance into float.
data=data[data['trip_distance']!=0] #removing trips with trip_distance=0.

data.shape[0] #now, the dataset is made of 6272091 rows.

data=data[data['total_amount']>=0] #removing trips where total_amount is negative (refunds).

data.shape[0] #now, the dataset is made of 6254985 rows.

6254985

Now we have the final dataset. We create the column *tripID* to enumerate the trips. 

In [97]:
data["tripID"]=range(1,len(data)+1) #creating column tripID.

To ease the computation, we'll split the dataset in two smaller dataset on *VendorID* and then join them again later.

In [99]:
V1=data[data['VendorID']==1] #dataset V1 is made of 2058650 rows.
V2=data[data['VendorID']==2] #dataset V2 is made of 4196335 rows.

This function, for each trip of the input dataset, returns all trips that can be chained to it into a list. The list is then saved into the column *chains_to*.

In [101]:
def find_chains_to(dataset):
    
    dataset['chains_to']=[[] for _ in range(len(dataset))] #creating the column chains_to.
    
    dataset=dataset.sort_values(by='tpep_pickup_datetime') #sorting the dataset by tpep_pickup_datetime, so that we can consider only the rows under the selected row.

    grouped=dataset.groupby('PULocationID') #grouping the dataset by PULocationID.
    datasets={location: group.reset_index(drop=True) for location, group in grouped} #saving each group as a new dataset, to ease the computing later.
    
    chains_to_list=[] #creating a list to hold the results.

    for i in range(len(dataset)-1): #each row checks only the rows under it, since the dataset is sorted by tpep_pickup_datetime.
        row=dataset.iloc[i] #saving the selected row.

        if row["DOLocationID"] in datasets: #checking if there are other trips from the row's dropoff location.
            
            c1=datasets[row["DOLocationID"]]['tpep_pickup_datetime']>row['tpep_dropoff_datetime'] #checking that the first trip ends before the second trip starts.
            c2=(datasets[row["DOLocationID"]]['tpep_pickup_datetime']-row['tpep_dropoff_datetime']).dt.total_seconds()/60<=2  #checking that the second trip starts after at most 2 minutes.

            possible_chains= (datasets[row["DOLocationID"]][c1&c2]['tripID']).tolist() #saving all tripIDs of the trips that sastisfy both conditions.
            chains_to_list.append(possible_chains) #adding the results to the main list.
        else:
            chains_to_list.append([]) #if there are no other trips from the row's dropoff location, we simply add an empty list.

    chains_to_list.append([]) #we don't check for the last trip since there are no trips that can be chained to it by default, so we must add an empty list at the end.

    dataset['chains_to']=chains_to_list #putting the results into the chains_to column in the dataset.

    return dataset #returning the updated dataset.

We'll run this function for both datasets and then put them back together.

In [103]:
V1=V1.reset_index(drop=True) #resetting the index.
V1=find_chains_to(V1)

V2=V2.reset_index(drop=True) #resetting the index.
V2=find_chains_to(V2)

data=pd.concat([V1,V2],ignore_index=True) #putting both datasets together again.

Let's see the dataset.

In [153]:
data.head(5)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0       1.0  2020-01-01 00:00:00   2020-01-01 00:13:03              1.0   
1       1.0  2020-01-01 00:00:03   2020-01-01 00:13:04              2.0   
2       1.0  2020-01-01 00:00:05   2020-01-01 00:26:30              0.0   
3       1.0  2020-01-01 00:00:07   2020-01-01 00:03:26              3.0   
4       1.0  2020-01-01 00:00:25   2020-01-01 00:05:59              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            2.2         1.0              False            68           170   
1            3.0         1.0              False            79           162   
2            3.4         1.0              False           264            68   
3            0.6         1.0              False            75            75   
4            1.7         1.0              False           145           179   

   payment_type  ...  total_amount  congestion_surcharge        duration  \
0           1.0  ...         17.15                   2.5 0 days 00:13:03   
1           1.0  ...         17.30                   2.5 0 days 00:13:01   
2           1.0  ...         26.15                   2.5 0 days 00:26:25   
3           2.0  ...          5.80                   0.0 0 days 00:03:19   
4           2.0  ...          8.30                   0.0 0 days 00:05:34   

   pickup_date  pickup_time  dropoff_date  dropoff_time     interval tripID  \
0   2020-01-01     00:00:00    2020-01-01      00:13:03  00:00-00:30   8387   
1   2020-01-01     00:00:03    2020-01-01      00:13:04  00:00-00:30   3044   
2   2020-01-01     00:00:05    2020-01-01      00:26:30  00:00-00:30   5694   
3   2020-01-01     00:00:07    2020-01-01      00:03:26  00:00-00:30   4990   
4   2020-01-01     00:00:25    2020-01-01      00:05:59  00:00-00:30   2522   

                                     chains_to  
0  [3540, 12318, 10968, 660, 3329, 8164, 2542]  
1                                           []  
2                   [8965, 2590, 11353, 12011]  
3                                           []  
4                                           []  

[5 rows x 26 columns]

Last, we thought to define a function that, given a trip, would return one of the possible chains that can be built starting from it. It will be a recursive function, that breaks when it faces a trip with no possible chains to it.

In [148]:
def build_chains(ID,data):
    print(ID) #printing current ID.
    if data[data["tripID"]==ID]["chains_to"].iloc[0]==[]: #checking if the current trip can be chained to other trips.
        return #if the current trip can't be chained, the function stops.
    else:
        build_chains(data[data["tripID"]==ID]["chains_to"].iloc[0][0],data) #repeating the process on the first of those trips that can be chained.

Here's an example for trip *8387*.

In [150]:
build_chains(8387,data) #returning 8387,3540,7351.

8387
3540
7351
